---
title: "Using R & Bioconductor to assign cell types to single-cell RNA-seq data"
output: 
  html_document:
    toc: true
    toc_float: true
author: Kieran R Campbell (kicampbell@bccrc.ca)
date: "`r format(Sys.time(), '%d %B, %Y')`"

---

# Overview of this tutorial

After going through this tutorial you should be able to:

1. Read 10X single-cell RNA-seq data into `SingleCellExperiment` objects
2. Know how to navigate `SingleCellExperiment` objects
3. Basic QC of single-cell RNA-seq data using `scater`
4. Create low dimensional plots (PCA, UMAP, tSNE)
5. Assign cells to known cell types using `cellassign`


## Resources

* [SingleCellExperiment vignette](https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html)
* [Scater vignette](https://bioconductor.org/packages/devel/bioc/vignettes/scater/inst/doc/vignette-intro.html)
* [CellAssign R package](https://github.com/irrationone/cellassign)
* [CellAssign preprint](http://bit.ly/cellassignpreprint)
* [Hemberg lab scRNA-seq course](https://hemberg-lab.github.io/scRNA.seq.course/index.html)
* [Orchestrating Single-Cell Analysis with Bioconductor](http://osca.bioconductor.org/)

**This document compiled:** http://kieranrcampbell.github.io/r-workshop-march-2019

## Required packages

```{r cache = FALSE}
suppressPackageStartupMessages({
  library(scater) # BioConductor
  library(SingleCellExperiment) # BioConductor
  library(DropletUtils) # BioConductor
  library(tidyverse) # CRAN
  library(here) # CRAN
  library(DT) # CRAN
  library(pheatmap) # CRAN
})

knitr::opts_chunk$set(echo = TRUE,
                      cache = TRUE)
```

If you need to install any of these, you can do so by:

```{r eval=F}
install.packages(c("tidyverse", "here", "DT", "pheatmap", "BiocManager"))

BiocManager::install(c("scater", "SingleCellExperiment", "DropletUtils", "biomaRt"))
```

**It is highly recommended you do so before the tutorial as this can take significant time!**

# Installation of cellassign

To install cellassign, we need to install Google's Tensorflow framework. **If this doesn't work in the tutorial - don't worry, you won't need it for 80% of what we cover**.

```{r, eval=T}
install.packages("tensorflow", repos = "http://cran.us.r-project.org")
tensorflow::install_tensorflow()

install.packages("devtools", repos = "http://cran.us.r-project.org") # If not already installed
devtools::install_github("Irrationone/cellassign")

```


# Read in 10X scRNA-seq data

We're going to use some re-processed data from [Single cell RNA sequencing of human liver reveals distinct intrahepatic macrophage populations](https://www.nature.com/articles/s41467-018-06318-7) (patient 1 specifically). If you git cloned this repository (https://github.com/kieranrcampbell/r-workshop-march-2019) this can be found in the directory `data/outs/filtered_gene_bc_matrices/GRCh38`:

```{r}
data_dir <- here("data", "outs", "filtered_gene_bc_matrices", "GRCh38")

print(data_dir)

print(dir(data_dir))
```

10X data typically has a barcodes file (indicating cell barcodes), a genes file (indicating a mapping between genes and indices) and the actual expression data as **raw counts** in `matrix.mtx`.

We can read these in to a `SingleCellExperiment` object using the `read10xCounts` function

```{r}
sce <- read10xCounts(data_dir)

sce
```

# A brief tour of the `SingleCellExperiment` package

Overall idea: think of your count matrix where the **columns are cells** and the **rows are genes**.

So when you see things like `rowData` think `geneData`, and `colData` think `cellData`!





## Data dimensions

Getting data dimensions

```{r}
nrow(sce) # Number of rows = genes
ncol(sce) # Number of columns = cells
```

```{r}
print(sce)
```


## Subsetting

```{r}
sce[, c(1,3,5)] # Subset to cells 1, 3, 5
```

```{r}
sce[c(2,4,6), ] # Subset to genes 2, 4, 6 
```

## Feature and cell metadata

`rownames` = gene names

```{r}
head(rownames(sce))
```

`colnames` = cell names = barcodes (sometimes)

```{r}
head(colnames(sce))
```




Column data (cell specific metadata)

```{r}
head(colData(sce))
```

I might want to set the column names to the barcode:

```{r}
colnames(sce) <- colData(sce)$Barcode

head(colnames(sce))
```

And I can subset on barcode:

```{r}
sce[, "AAACCTGCAGTAAGCG-1"]
```



Row data (gene specific metadata)

```{r}
head(rowData(sce))
```


`reducedDims` is where our PCA,UMAP,tSNE representations will live - but we haven't made them yet

```{r}
reducedDims(sce)
```


`sizeFactors` is where our cell size factors will live - but we haven't calculated them yet

```{r}
head(sizeFactors(sce))
```


The ability to have multiple `assays` is one of the unique advantages of the SingleCellExperiment approach - I can carry around `counts`, `logcounts`, and any other weird data transformation I might like. Right now we only have raw counts, because that's what we've read in:

```{r}
names(assays(sce))
```

```{r, eval = FALSE}
assay(sce, "counts")
counts(sce)
```


I can make my own:

```{r}
assay(sce, "my_super_strange_assay") <- sin(as.matrix(counts(sce)))

names(assays(sce))
```

```{r}
class(assay(sce, "my_super_strange_assay"))
```

Note the beauty of SingleCellExperiments is that subsetting is consistent: if I want to subset only some cells and genes:

```{r}
sce_subset <- sce[c(1,3,5), c(2,4,6,8)]
```

Then everything else is subset too!

```{r}
print(dim(counts(sce_subset)))
print(length(sizeFactors(sce_subset)))
print(dim(rowData(sce_subset)))
```

So the approach may seem like a lot of work up front to just carry around a count matrix and some metadata, but this sort of consistent subsetting makes it _much_ harder (but still not impossible) to introduce bugs into your analysis.



# Quality control of scRNA-seq data

## Getting started

First we do some key things to our data:

1. Get some extra gene data, including the chromosome name
2. Compute the size factors
3. Compute normalized log counts 

```{r}
rowData(sce)$ensembl_gene_id <- rownames(sce)
  
sce <- getBMFeatureAnnos(sce, 
                         filters = "ensembl_gene_id",
                         attributes = c("ensembl_gene_id", "hgnc_symbol", "entrezgene",
                                        "start_position", "end_position", "chromosome_name"),
                         dataset = "hsapiens_gene_ensembl")

  
# Calculate size factors
sce <- scran::computeSumFactors(sce, BPPARAM = MulticoreParam(10))


# Compute log normal expression values
sce <- normalize(sce)

```

```{r}
names(assays(sce))
```

```{r}
head(sizeFactors(sce))
```


```{r}
sce <- runPCA(sce)
sce <- runUMAP(sce)
```

```{r}
reducedDims(sce)
```

Just give me my PCA!

```{r}
head(reducedDims(sce)[['PCA']])
```


I like to add the symbols to the rownames:

```{r}
rownames(sce) <- paste0(rowData(sce)$Symbol, "_", rownames(sce))
head(rownames(sce))
```


We next need to work out which genes are mitochondrial and ribosomal as these work well for QC:

```{r}
# Get Mitochondrial genes for QC:
mt_genes <- which(rowData(sce)$chromosome_name == "MT")
ribo_genes <- grepl("^RP[LS]", rowData(sce)$Symbol)
feature_ctrls <- list(mito = rownames(sce)[mt_genes],
                      ribo = rownames(sce)[ribo_genes])

lapply(feature_ctrls, head)
```

And call the `calcualteQCMetrics` function in scater:

```{r}
sce <- calculateQCMetrics(sce, feature_controls = feature_ctrls)
```

```{r}
datatable(head(as.data.frame(colData(sce))))
```


## What to look for

My personal favourite plot to QC scRNA-seq data:

```{r}
plotColData(sce, x = "total_features_by_counts", y = "pct_counts_mito")
```

Typically retain cells that have < 10% mitochondrial transcripts and > 1000 features, but this is **dataset dependent** - for example, tumour cells typically have higher metabolic burden, leading to higher % mitochondrial (we typically use 20% as a filter then).

```{r}
plotPCA(sce, colour_by = "pct_counts_mito")
```

For this going to use a simple threshold of 10% mitochondrial. Importantly, we re-compute the QC metrics, size factors and normalization

```{r}
mito_thresh <- 10

sce_qc <- sce[, sce$pct_counts_mito < mito_thresh]

sce_qc <- scran::computeSumFactors(sce_qc, BPPARAM = MulticoreParam(10))

sce_qc <- normalize(sce_qc)

sce_qc <- calculateQCMetrics(sce_qc, feature_controls = feature_ctrls)
```

```{r}
sce_qc <- runPCA(sce_qc)
sce_qc <- runUMAP(sce_qc)
```

```{r}
# plotPCA(sce_qc, colour_by = "total_features_by_counts")
plotUMAP(sce_qc, colour_by = "pct_counts_mito")
```

```{r}
plotScater(sce)
plotScater(sce_qc)
```


```{r}
plotHighestExprs(sce_qc)
```

## Some useful plotting functions in Scater

### Reduced dimension plots

I can call 

* `plotPCA(sce, colour_by = "x")`
* `plotUMAP(sce, colour_by = "x")`
* `plotTSNE(sce, colour_by = "x")`

where `x` is:

* Any column of `colData(sce)` (= the cell specific data) to colour by metadata
* Any gene name in `rownames(sce)` to colour by expression

```{r}
plotPCA(sce, colour_by = "SAA1_ENSG00000173432")
```

### Additional plots

```{r}
plotColData(sce_qc, x = "total_counts", y = "pct_counts_mito")
```

```{r}
plotExpression(sce_qc, 
               x = "total_counts", 
               features = "GAPDH_ENSG00000111640")
```



# Using CellAssign to assign cells to known types

CellAssign is our new method to assign cells to known cell types. It relies on assuming cell types over-express their own markers, e.g. an epithelial tumour cell should overexpress EPCAM relative to other cell types. 

```{r}
library(cellassign)
```



In this example, the data we have just performed QC and exploratory analysis of is liver cells, that we expect to contain a certain number of cell types. To begin, we specify a list, where each item corresponds to a set of marker genes for a given cell type:

```{r}
liver_marker_list <- list(
        Hepatocytes = c("ALB", "HAMP", "ARG1", "PCK1", "AFP", "BCHE"), 
        LSECs = c("CALCRL", "FCGR2B", "VWF"),
        Cholangiocytes = c("KRT19", "EPCAM", "CLDN4", "CLDN10", "SOX9", "MMP7", "CXCL1", "CFTR", "TFF2", "KRT7", "CD24"), 
        `Hepatic Stellate Cells` = c("ACTA2", "COL1A1", "COL1A2", "COL3A1", "DCN", "MYL9"),
        Macrophages = c("CD68", "MARCO", "FCGR3A", "LYZ", "PTPRC"),
        `ab T cells` = c("CD2", "CD3D", "TRAC", "IL32", "CD3E", "PTPRC"),
        `gd T cells` = c("NKG7", "FCGR3A", "HOPX", "GNLY", "KLRF1", "CMC1", "CCL3", "PTPRC"),
        `NK cells` = c("GZMK", "KLRF1", "CCL3", "CMC1", "NKG7", "PTPRC"),
        `Plasma cells` = c("CD27", "IGHG1", "CD79A", "IGHG2", "PTPRC", "IGKC"),
        `Mature B cells` = c("MS4A1", "LTB", "CD52", "IGHD", "CD79A", "PTPRC", "IGKC"),
        `Erythroid cells` = c("HBB", "SLC25A37", "CA1", "ALAS2")
)
```

To begin, we use `cellassign`'s `marker_list_to_mat` function to convert this into a (binary) cell type by marker matrix:

```{r, fig.width = 5, fig.height = 10}
mgi <- marker_list_to_mat(liver_marker_list, include_other = FALSE)

pheatmap(mgi)
```

We then make sure all of these markers exist in our SingleCellExperiment:

```{r}
marker_in_sce <- match(rownames(mgi), rowData(sce_qc)$Symbol)
stopifnot(all(!is.na(marker_in_sce)))
```

And finally we subset `sce` to just the marker genes:

```{r}
sce_marker <- sce_qc[marker_in_sce, ]
```

```{r}
stopifnot(all.equal(rownames(mgi), rowData(sce_marker)$Symbol))
```

We then call `cellassign` passing in the `SingleCellExperiment`, marker info, the size factors we've calculated, as well as various other options:


```{r warning = F, message = F}
counts(sce_marker) <- as.matrix(counts(sce_marker))

print(dim(sce_marker))
print(dim(mgi))

fit <- cellassign(
  exprs_obj = sce_marker,
  marker_gene_info = mgi,
  s = sizeFactors(sce_qc),
  shrinkage = TRUE,
  max_iter_adam = 50,
  min_delta = 2,
  verbose = TRUE
)

```



```{r}
fit

```


Add the cell types to the sce:

```{r}
sce_qc$cell_type <- fit$cell_type

```

```{r}
plotUMAP(sce_qc, colour_by = "cell_type")

```


```{r, fig.width = 9, fig.height = 10}
acol <- data.frame(`cellassign cell type` = sce_qc$cell_type)
rownames(acol) <- colnames(sce_qc)

pheatmap(as.matrix(logcounts(sce_marker)),
         annotation_col = acol)

```

```{r}
pheatmap(fit$mle_params$gamma)

```



